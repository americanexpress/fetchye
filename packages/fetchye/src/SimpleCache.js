/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  IS_LOADING,
  SET_DATA,
  DELETE_DATA,
  ERROR,
  CLEAR_ERROR,
  ACTION_NAMESPACE,

} from 'fetchye-core';

// eslint-disable-next-line default-param-last -- the first default param value takes care of explicitly calling this function with `undefined` the second param can't be defaulted as it must be provided
function reducer(state = {
  errors: {},
  loading: {},
  data: {},
}, action) {
  if (!action.type.startsWith(ACTION_NAMESPACE)) {
    return state;
  }
  switch (action.type) {
    case DELETE_DATA: {
      const { [action.hash]: deletedEntry, ...nextData } = state.data;
      return {
        ...state,
        data: {
          ...nextData,
        },
      };
    }
    case CLEAR_ERROR: {
      const { [action.hash]: deletedEntry, ...nextErrors } = state.errors;
      return {
        ...state,
        errors: {
          ...nextErrors,
        },
      };
    }
    case ERROR: {
      const { [action.hash]: deletedEntry, ...nextLoading } = state.loading;
      return {
        ...state,
        errors: {
          ...state.errors,
          [action.hash]: action.error,
        },
        loading: {
          ...nextLoading,
        },
      };
    }
    case IS_LOADING: {
      return {
        ...state,
        loading: {
          ...state.loading,
          [action.hash]: action.hash,
        },
      };
    }
    case SET_DATA: {
      const { [action.hash]: deletedEntry, ...nextLoading } = state.loading;
      return {
        ...state,
        data: {
          ...state.data,
          [action.hash]: action.value,
        },
        loading: {
          ...nextLoading,
        },
      };
    }
    default:
      return state;
  }
}

const getCacheByKey = (cache = {}, key = undefined) => {
  const data = cache.data?.[key];
  const loading = !!cache.loading?.[key];
  const error = cache.errors?.[key];
  return { data, loading, error };
};

function SimpleCache({ cacheSelector = (state) => state } = {}) {
  return {
    getCacheByKey,
    reducer,
    cacheSelector,
  };
}
export default SimpleCache;
