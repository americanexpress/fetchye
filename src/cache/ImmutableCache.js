/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// Immutable is an optional dependency
// eslint-disable-next-line import/no-extraneous-dependencies
import { Map as iMap } from 'immutable';
import reducer from './immutable/reducer';

const getCacheByKey = (cache = iMap(), key) => {
  const data = cache.getIn(['data', key]);
  const loading = cache.hasIn(['loading', key]);
  const error = cache.getIn(['errors', key]);
  return { data, loading, error };
};

export default function ImmutableCache({ cacheSelector = (state) => state } = {}) {
  return {
    getCacheByKey,
    reducer,
    cacheSelector,
  };
}
