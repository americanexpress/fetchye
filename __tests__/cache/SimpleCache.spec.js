/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { createStore } from 'redux';
import SimpleCache from '../../src/cache/SimpleCache';
import * as actions from '../../src/cache/actions';

const fakeError = new Error('Fake Error');
const fakeData = {
  status: 200,
  ok: true,
  body: {
    fakeData: true,
  },
};

const createScenario = (dispatch, actionKeys, hash) => {
  const actionOrder = actionKeys.map((key) => actions[key]);
  actionOrder.forEach((actionCreator) => {
    dispatch(
      actionCreator({
        hash,
        error: fakeError,
        value: fakeData,
      })
    );
  });
};

describe('SimpleCache', () => {
  let store;
  const cache = SimpleCache();
  beforeEach(() => {
    store = createStore(cache.reducer, cache.reducer(undefined, { type: '' }));
  });
  it('should reflect load state', () => {
    const { dispatch, getState } = store;
    createScenario(dispatch, ['loadingAction'], 'abc1234');
    expect(getState()).toMatchInlineSnapshot(`
      Object {
        "data": Object {},
        "errors": Object {},
        "loading": Object {
          "abc1234": "abc1234",
        },
      }
    `);
  });
  it('should reflect load to success state', () => {
    const { dispatch, getState } = store;
    createScenario(dispatch, ['loadingAction', 'setAction'], 'abc1234');
    expect(getState()).toMatchInlineSnapshot(`
      Object {
        "data": Object {
          "abc1234": Object {
            "body": Object {
              "fakeData": true,
            },
            "ok": true,
            "status": 200,
          },
        },
        "errors": Object {},
        "loading": Object {},
      }
    `);
  });
  it('should reflect load to error state', () => {
    const { dispatch, getState } = store;
    createScenario(dispatch, ['loadingAction', 'errorAction'], 'abc1234');
    expect(getState()).toMatchInlineSnapshot(`
      Object {
        "data": Object {},
        "errors": Object {
          "abc1234": [Error: Fake Error],
        },
        "loading": Object {},
      }
    `);
  });
  it('should reflect load to success to delete state', () => {
    const { dispatch, getState } = store;
    createScenario(dispatch, ['loadingAction', 'setAction', 'deleteAction'], 'abc1234');
    expect(getState()).toMatchInlineSnapshot(`
      Object {
        "data": Object {},
        "errors": Object {},
        "loading": Object {},
      }
    `);
  });
  it('should reflect load to error to clear errors state', () => {
    const { dispatch, getState } = store;
    createScenario(dispatch, ['loadingAction', 'errorAction', 'clearErrorsAction'], 'abc1234');
    expect(getState()).toMatchInlineSnapshot(`
      Object {
        "data": Object {},
        "errors": Object {},
        "loading": Object {},
      }
    `);
  });
  it('should reflect multiple hashes stored', () => {
    const { dispatch, getState } = store;
    createScenario(dispatch, ['loadingAction', 'setAction'], 'abc1234');
    createScenario(dispatch, ['loadingAction', 'errorAction'], 'def5678');
    expect(getState()).toMatchInlineSnapshot(`
      Object {
        "data": Object {
          "abc1234": Object {
            "body": Object {
              "fakeData": true,
            },
            "ok": true,
            "status": 200,
          },
        },
        "errors": Object {
          "def5678": [Error: Fake Error],
        },
        "loading": Object {},
      }
    `);
  });
  it('should return default state if unknown @fetchye action type', () => {
    const { dispatch, getState } = store;
    dispatch({ type: '@fetchye' });
    expect(getState()).toMatchInlineSnapshot(`
      Object {
        "data": Object {},
        "errors": Object {},
        "loading": Object {},
      }
    `);
  });

  describe('getCacheByKey', () => {
    it('should return data error loading', () => {
      const { dispatch, getState } = store;
      createScenario(dispatch, ['loadingAction', 'setAction'], 'abc1234');
      createScenario(dispatch, ['loadingAction', 'errorAction'], 'def5678');
      expect(cache.getCacheByKey(getState(), 'abc1234')).toMatchInlineSnapshot(`
        Object {
          "data": Object {
            "body": Object {
              "fakeData": true,
            },
            "ok": true,
            "status": 200,
          },
          "error": undefined,
          "loading": undefined,
        }
      `);
    });
    it('should return empty data error loading if cache undefined', () => {
      expect(cache.getCacheByKey(undefined, 'abc1234')).toMatchInlineSnapshot(`
        Object {
          "data": undefined,
          "error": undefined,
          "loading": undefined,
        }
      `);
    });
  });
  it('should accept a cacheSelector', () => {
    const { dispatch, getState } = store;
    createScenario(dispatch, ['loadingAction', 'errorAction'], 'def5678');
    const cacheSelector = (state) => state.someSliceOfState;
    const nextCache = SimpleCache({ cacheSelector });
    expect(nextCache.cacheSelector({ someSliceOfState: getState() })).toMatchInlineSnapshot(`
        Object {
          "data": Object {},
          "errors": Object {
            "def5678": [Error: Fake Error],
          },
          "loading": Object {},
        }
      `);
  });
  it('should return default cacheSelector', () => {
    const { dispatch, getState } = store;
    createScenario(dispatch, ['loadingAction', 'errorAction'], 'def5678');
    const nextCache = SimpleCache();
    expect(nextCache.cacheSelector(getState())).toMatchInlineSnapshot(`
      Object {
        "data": Object {},
        "errors": Object {
          "def5678": [Error: Fake Error],
        },
        "loading": Object {},
      }
    `);
  });
});
