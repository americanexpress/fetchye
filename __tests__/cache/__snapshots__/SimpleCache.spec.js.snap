// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleCache accepts a cacheSelector 1`] = `
Object {
  "data": Object {},
  "errors": Object {
    "def5678": [Error: Fake Error],
  },
  "loading": Object {},
}
`;

exports[`SimpleCache getCacheByKey returns empty data error loading if cache undefined 1`] = `
Object {
  "data": undefined,
  "error": undefined,
  "loading": false,
}
`;

exports[`SimpleCache reflects load state 1`] = `
Object {
  "data": Object {},
  "errors": Object {},
  "loading": Object {
    "abc1234": "abc1234",
  },
}
`;

exports[`SimpleCache reflects load to error state 1`] = `
Object {
  "data": Object {},
  "errors": Object {
    "abc1234": [Error: Fake Error],
  },
  "loading": Object {},
}
`;

exports[`SimpleCache reflects load to error to clear errors state 1`] = `
Object {
  "data": Object {},
  "errors": Object {},
  "loading": Object {},
}
`;

exports[`SimpleCache reflects load to success state 1`] = `
Object {
  "data": Object {
    "abc1234": Object {
      "body": Object {
        "fakeData": true,
      },
      "ok": true,
      "status": 200,
    },
  },
  "errors": Object {},
  "loading": Object {},
}
`;

exports[`SimpleCache reflects load to success to delete state 1`] = `
Object {
  "data": Object {},
  "errors": Object {},
  "loading": Object {},
}
`;

exports[`SimpleCache reflects multiple hashes stored 1`] = `
Object {
  "data": Object {
    "abc1234": Object {
      "body": Object {
        "fakeData": true,
      },
      "ok": true,
      "status": 200,
    },
  },
  "errors": Object {
    "def5678": [Error: Fake Error],
  },
  "loading": Object {},
}
`;

exports[`SimpleCache returns default cacheSelector 1`] = `
Object {
  "data": Object {},
  "errors": Object {
    "def5678": [Error: Fake Error],
  },
  "loading": Object {},
}
`;

exports[`SimpleCache returns default state if unknown @fetchye action type 1`] = `
Object {
  "data": Object {},
  "errors": Object {},
  "loading": Object {},
}
`;
